File: app/src/main/java/io/ffem/integration/MainBaseActivity.java
package io.ffem.integration;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.Spinner;
import android.widget.TextView;

import java.util.ArrayList;

/**
 * This class contains just the setup for the demo app only
 * This code is not required for integration.
 * Please see MainActivity for the integration code.
 */
public class MainBaseActivity extends BaseActivity {

    TextView textResult;
    CheckBox dummyResultCheckBox;
    String selectedTest;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setAppTheme();

        setContentView(R.layout.activity_main);

        initialize();
    }

    void clearResultDisplay() {
        textResult.setText("");
    }

    private void initialize() {
        textResult = findViewById(R.id.result);
        dummyResultCheckBox = findViewById(R.id.dummyResultCheckBox);
        dummyResultCheckBox.setOnCheckedChangeListener((compoundButton, b) -> clearResultDisplay());
        Spinner spinner = findViewById(R.id.spinner);

        ArrayList<String> tests = new ArrayList<>();
        tests.add(Constants.AVAILABLE_IRON);
        tests.add(Constants.CALCIUM_MAGNESIUM);
        tests.add(Constants.FLUORIDE);
        tests.add(Constants.INVALID_TEST);

        // Creating adapter for spinner
        ArrayAdapter<String> dataAdapter = new ArrayAdapter<>(this,
                android.R.layout.simple_spinner_item, tests);

        // Drop down layout style - list view with radio button
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        // attaching data adapter to spinner
        spinner.setAdapter(dataAdapter);

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                clearResultDisplay();
                selectedTest = parent.getItemAtPosition(position).toString();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
    }

    public interface DialogDismissListener {
        void dialogDismissed(DialogInterface dialog);
    }
}

File: app/src/main/java/io/ffem/integration/MainBaseActivity.kt
package io.ffem.integration

import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.CompoundButton
import androidx.appcompat.app.AlertDialog
import io.ffem.integration.PreferencesUtil.getInt
import io.ffem.integration.PreferencesUtil.setInt
import io.ffem.integration.databinding.ActivityMainBinding
import java.util.*

/**
 * This class contains just the setup for the demo app only
 * This code is not required for integration.
 * Please see MainActivity for the integration code.
 */
open class MainBaseActivity : BaseActivity() {
    protected lateinit var b: ActivityMainBinding
    var selectedTest: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        b = ActivityMainBinding.inflate(layoutInflater)
        val view = b.root
        setContentView(view)
        setAppTheme()
        initialize()
    }

    fun clearResultDisplay() {
        b.textResult.text = ""
    }

    private fun initialize() {
        b.checkDummyResult.setOnCheckedChangeListener { _: CompoundButton?, _: Boolean -> clearResultDisplay() }

        val tests = mutableListOf(
            AVAILABLE_IRON,
            CALCIUM_MAGNESIUM,
            WATER_CALCIUM_MAGNESIUM,
            FLUORIDE,
            FLUORIDE_LITE,
            RESIDUAL_CHLORINE_LITE,
            TOTAL_ALKALINITY,
            TOTAL_HARDNESS,
            INVALID_TEST
        )

        // Creating adapter for spinner
        val dataAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, tests)

        // Drop down layout style - list view with radio button
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)

        // attaching data adapter to spinner
        b.spinnerList.adapter = dataAdapter
        b.spinnerList.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>,
                view: View,
                position: Int,
                id: Long
            ) {
                clearResultDisplay()
                selectedTest = parent.getItemAtPosition(position).toString()

                setInt(baseContext, R.string.selected_test_key, position)
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        // set the selected test to previously selected one
        b.spinnerList.setSelection(getInt(baseContext, R.string.selected_test_key, 0))
    }

    protected fun showAppNotInstalledDialog(appTitle: String?, externalAppAction: String?) {
        val builder = AlertDialog.Builder(this, R.style.Theme_MaterialComponents_Light_Dialog_Alert)
        builder.setTitle(R.string.app_not_found)
            .setMessage(String.format(Locale.US, getString(R.string.install_app), appTitle))
            .setPositiveButton(R.string.go_to_play_store) { _: DialogInterface?, _: Int ->
                startActivity(
                    Intent(
                        Intent.ACTION_VIEW,
                        Uri.parse(PLAY_STORE_URL + externalAppAction)
                    )
                )
            }
            .setNegativeButton(
                android.R.string.cancel
            ) { dialogInterface: DialogInterface, _: Int -> dialogInterface.dismiss() }
            .setCancelable(false)
            .show()
    }
}